File: iOS/AgoraEntScenarios/Scenes/SBG/SBGAPI/LyricsCutter.swift
//
//  LyricsCutter.swift
//  Demo
//
//  Created by ZYP on 2023/4/25.
//

import Foundation
import AgoraLyricsScore

class LyricsCutter {
    /// 每一行歌词的信息
    public struct Line {
        var beginTime: Int
        var duration: Int
        var endTime: Int {
            return beginTime + duration
        }
    }
    
    /// 处理时间副歌片段时间（对齐句子）
    /// - Parameters:
    ///   - startTime: 副歌开始时间
    ///   - endTime: 副歌结束时间
    ///   - lines: 歌词所有行数据
    /// - Returns: (startTime, endTime), nil为无法处理
    public static func handleFixTime(startTime: Int,
                                     endTime: Int,
                                     lines: [Line]) -> (Int, Int)? {
        guard startTime < endTime else {
            return nil
        }
        
        guard startTime != endTime else {
            return nil
        }
        
        guard !lines.isEmpty else {
            return nil
        }
        
        var start = startTime
        var end = endTime
        
        if start < lines.first!.beginTime,
            end < lines.first!.beginTime {
            return nil
        }
        
        if start > lines.last!.endTime,
           end > lines.last!.endTime {
            return nil
        }
        
        /// 跨过第一个
        if start < lines.first!.beginTime, end < lines.first!.endTime {
            start = lines.first!.beginTime
            end = lines.first!.endTime
            return (start, end)
        }
        
        /// 跨过最后一个
        if start > lines.last!.beginTime,
           end > lines.last!.endTime {
            start = lines.last!.beginTime
            end = lines.last!.endTime
            return (start, end)
        }
        
        if start < lines.first!.beginTime {
            start = lines.first!.beginTime
        }
        
        if end > lines.last!.endTime {
            end = lines.last!.endTime
        }
        
        var startIndex = 0
        var startGap = Int.max
        var endIndex = 0
        var endGap = Int.max
        for (offset, line) in lines.enumerated() {
            if abs(line.beginTime - start) < startGap {
                startGap = abs(line.beginTime - start)
                startIndex = offset
            }
            if abs(line.endTime - end) < endGap {
                endGap = abs(line.endTime - end)
                endIndex = offset
            }
        }
        
        let result = (lines[startIndex].beginTime, lines[endIndex].endTime)
        if result.0 < result.1 { /** valid **/
            return result
        }
        return nil
    }
    
    /// 裁剪副歌片段
    /// - Parameters:
    ///   - model: 歌词模型
    ///   - start: 副歌开始时间
    ///   - end: 副歌结束时间
    /// - Returns: 副歌模型
    public static func cut(model: LyricModel, startTime: Int, endTime: Int) -> LyricModel {
        var lines = [LyricLineModel]()
        
        var flag = false
        for line in model.lines {
            if line.beginTime == startTime {
                flag = true
            }
            if line.beginTime + line.duration == endTime {
                lines.append(line)
                break
            }
            if flag {
                lines.append(line)
            }
        }
        
        model.lines = lines
        model.preludeEndPosition = 0
        model.duration = (lines.last?.beginTime ?? 0) + (lines.last?.duration ?? 0)
        return model
    }
}

File: Android/scenes/ktv_singbattle/src/main/java/io/agora/scene/ktv/singbattle/widget/lrcView/LyricsCutter.java
package io.agora.scene.ktv.singbattle.widget.lrcView;

import android.util.Pair;

import java.util.ArrayList;
import java.util.List;

import io.agora.karaoke_view_ex.internal.model.LyricsLineModel;
import io.agora.karaoke_view_ex.model.LyricModel;
import io.agora.scene.ktv.singbattle.KTVLogger;

/**
 * 歌曲裁剪
 */
public class LyricsCutter {

    static class Line {
        private int beginTime;
        private int duration;

        public Line(long beginTime, long duration) {
            this.beginTime = (int) beginTime;
            this.duration = (int) duration;
        }

        public int getBeginTime() {
            return beginTime;
        }

        public int getDuration() {
            return duration;
        }

        public int getEndTime() {
            return beginTime + duration;
        }
    }

    private static String tag = "LyricsCutter";

    // 处理时间副歌片段时间（对齐句子）
    public static Pair<Integer, Integer> handleFixTime(int startTime, int endTime, List<Line> lines) {
        if (startTime >= endTime || lines.isEmpty()) {
            return null;
        }

        int start = startTime;
        int end = endTime;

        Line firstLine = lines.get(0);
        Line lastLine = lines.get(lines.size() - 1);

        if ((start < firstLine.getBeginTime() && end < firstLine.getBeginTime()) ||
                (start > lastLine.getEndTime() && end > lastLine.getEndTime())) {
            return null;
        }

        // 跨过第一个
        if (start < firstLine.getBeginTime() && end < firstLine.getEndTime()) {
            start = firstLine.getBeginTime();
            end = firstLine.getEndTime();
            return new Pair<>(start, end);
        }
        // 跨过最后一个
        if (start > lastLine.getBeginTime() && end > lastLine.getEndTime()) {
            start = lastLine.getBeginTime();
            end = lastLine.getEndTime();
            return new Pair<>(start, end);
        }

        if (start < firstLine.getBeginTime()) {
            start = firstLine.getBeginTime();
        }
        if (end > lastLine.getEndTime()) {
            end = lastLine.getEndTime();
        }

        int startIndex = 0;
        int startGap = Integer.MAX_VALUE;
        int endIndex = 0;
        int endGap = Integer.MAX_VALUE;

        for (int i = 0; i < lines.size(); i++) {
            Line line = lines.get(i);
            int currentStartGap = Math.abs(line.getBeginTime() - start);
            int currentEndGap = Math.abs(line.getEndTime() - end);

            if (currentStartGap < startGap) {
                startGap = currentStartGap;
                startIndex = i;
            }
            if (currentEndGap < endGap) {
                endGap = currentEndGap;
                endIndex = i;
            }
        }

        Line startLine = lines.get(startIndex);
        Line endLine = lines.get(endIndex);
        if (startLine.getBeginTime() < endLine.getEndTime()) {
            return new Pair<>(startLine.getBeginTime(), endLine.getEndTime());
        }
        return null;
    }

    // 裁剪副歌片段
    public static LyricModel cut(LyricModel model, int startTime, int endTime) {
        KTVLogger.d(tag, "cut, startTime:" + startTime + "endTime:" + endTime);
        List<LyricsLineModel> lines = new ArrayList<>();
        boolean flag = false;

        for (LyricsLineModel line : model.lines) {
            if (line.getStartTime() == startTime) {
                flag = true;
            }
            if (line.getEndTime() == endTime) {
                lines.add(line);
                break;
            }
            if (flag) {
                lines.add(line);
            }
        }

        model.lines = lines;
        if (lines.isEmpty()) {
            model.duration = 0;
        } else {
            LyricsLineModel lastLine = lines.get(lines.size() - 1);
            model.duration = lastLine.getEndTime()/* - lastLine.getStartTime()*/;
        }

        return model;
    }
}
